### PR TITLE
Limiting the number of ImageSources cached to address memory issue

### PR DESCRIPTION
**Pull request by jjw24**
**Problem:**
Memory usage can run up quite high for Flow, and can hit 300 mbs for myself if constantly performing searches. It does reduce back down to around 130 when querying is stopped.

_Note:_
This is a partial fix to the memory issue, it does not address it complete, and other factors that contribute to it are still been investigated.

This PR does not mean memory will reduce significantly, it is more focused on addressing the unnecessary high memory caused by improperly handling of image caching. So reduce image cache (ImageSources)'s memory's footprint is the main focus here. 

**How to reproduce:**
Open query, type 'a' (especially if Explorer hotkey is set as global '*') and you should get numerous results. Keep pressing down arrow on the keyboard and should see Flow.Launcher.exe in task manager grow to as high as 300mbs for myself. It may vary depending on the user and the application images that need to be displayed on the machine.

If unable to repro, change the power mode on Windows to 'Recommended' then try.

**Partial solution:**
Limiting the number of ImageSources cached to reduced the memory usage going over too high.
- Set maximum number of images cached to 50 from 5000.

- Once cached images go over permissible factor (2) X number of images (50), perform clean up by removing infrequently used ones from dictionary.

**Testing**
1. Take snapshots of the memory usage by using the Diagnostic tools of Visual studio (Debug -> Windows -> Diagnostic tools).
2. Take the first snapshot after all the dlls are loaded and the memory utilization is stabilized.
3. Take the next memory snapshot after many queries (don't have to execute an application). The aim is to just load many icons.
4. Compare the second snapshot by setting the first one as the baseline.
5. Search for memory size diff in bytes for Concurrent Dictionary with value ImageSource.
6. Compare this value to that obtained without the changes in this PR, to get the results as shown in the following images.

_Result_
- Before this PR change:
image

- With this PR:
image

**Related**: #128

### CODE REVIEW

**Comment by taooceros** [comment]
@bao-qian has introduced a way to lazy load image for resultlistview, should I add that to this pr?

File: Flow.Launcher.Infrastructure/Image/ImageCache.cs [hunk 1]
```diff
@@ -2,17 +2,18 @@
...
 
 namespace Flow.Launcher.Infrastructure.Image
 {
     [Serializable]
     public class ImageCache
     {
-        private const int MaxCached = 5000;
+        private const int MaxCached = 50;
         public ConcurrentDictionary<string, int> Usage = new ConcurrentDictionary<string, int>();
         private readonly ConcurrentDictionary<string, ImageSource> _data = new ConcurrentDictionary<string, ImageSource>();
```
Comment (taooceros) [inline Flow.Launcher.Infrastructure/Image/ImageCache.cs:15]: Maybe we should merge this two dictionary to one by creating another utility class, which will allow us to remove element easier?

**Comment by jjw24** [comment]
> @bao-qian has introduced a way to lazy load image for resultlistview, which is a quite simple change, should I add that to this pr?

Are you referring to the change in the other PR ?

**Comment by taooceros** [comment]
> > @bao-qian has introduced a way to lazy load image for resultlistview, which is a quite simple change, should I add that to this pr?
> 
> Are you referring to the change in the other PR ?

Yes, it is part of commit in #195

**Comment by jjw24** [comment]
> > > @bao-qian has introduced a way to lazy load image for resultlistview, which is a quite simple change, should I add that to this pr?
> > 
> > 
> > Are you referring to the change in the other PR ?
> 
> Yes, it is part of commit in #195

I think we can keep them separate and then just resolve any conflict depending which goes in first

**Comment by taooceros** [comment]
Sure

File: Flow.Launcher.Infrastructure/Image/ImageCache.cs (hunk 1 above)
Comment (jjw24) [reply Flow.Launcher.Infrastructure/Image/ImageCache.cs:15]: yeah, let me have a look

Comment (taooceros) [reply Flow.Launcher.Infrastructure/Image/ImageCache.cs:15]: I have finished that, if you want, I can add that to the pr.

Comment (jjw24) [reply Flow.Launcher.Infrastructure/Image/ImageCache.cs:15]: haha ok cool, yes please

**Comment by jjw24** [comment]
@taooceros new changes look good. Do you want to approve and merge it in?

**Comment by taooceros** [comment]
Sure!

**Review by taooceros — APPROVED** [review]

**Pull request merged**